### PR TITLE
fix: fixed react18 React.render deprecation notice for WCAdmin

### PR DESCRIPTION
**Pull request by rjchow**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Related to https://github.com/woocommerce/woocommerce/issues/46879

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install WC Admin and do a simple smoke test, any failures should be obvious and catastrophic. Affected areas:
- WC Admin React home page

image

- WC Admin Embedded React pages, e.g Analytics, Marketing

image

- WC Admin Settings slot fills. e.g the "Site Visibility" tab

image

- Customer Effort Score

image

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by ilyasfoo — APPROVED** [review]
Tests well, LGTM! 🚢

**Review by chihsuan — APPROVED** [review]
LGTM and tested well! 🚢

**Comment by adrianduffell** [comment]
Nice work on this, @rjchow! I guess merging this would be blocked by issue https://github.com/woocommerce/woocommerce/issues/46879? 🤔

**Comment by rjchow** [comment]
> Nice work on this, @rjchow! I guess merging this would be blocked by issue #46879? 🤔

@adrianduffell #48933 actually

**Comment by mattsherman** [comment]
> > Nice work on this, @rjchow! I guess merging this would be blocked by issue #46879? 🤔
> 
> @adrianduffell #48933 actually

I'm actively working on a fix for #48933

**Comment by mattsherman** [comment]
@rjchow The following PR has been merged to `trunk`, which I believe addresses the issue (#48933) with the new product editor preventing this PR from being merged:

- https://github.com/woocommerce/woocommerce/pull/49058

I'd suggest rebasing this PR against trunk and verifying that things are indeed fixed (I tested with this PR in a draft PR branch, #49117, and things appeared fixed).

**Comment by rjchow** [comment]
Hey @mattsherman ! I think it's still happening - did you test on jurassic.ninja? It doesn't show up locally. I tested both this branch rebased on trunk, as well as #49117 

image

**Comment by mattsherman** [comment]
> Hey @mattsherman ! I think it's still happening - did you test on jurassic.ninja? It doesn't show up locally. I tested both this branch rebased on trunk, as well as #49117
> 
> image

Hrmm.... yes, I tested on JN. I wonder if there are any settings/plugins for your site that are different when you set things up on JN to test? Or, maybe it is network speed?

I'm AFK today, but I'll dig further into this on Monday!

**Comment by rjchow** [comment]
@mattsherman that's curious... I literally just launch a JN site from this PR page, 
- skip the core profiler, 
- set my country to US:CA, 
- enable product editor beta
- go to add product, variations
- add variation 
- see the block error

enjoy your AFK, let's look into this next week 😬

**Comment by mattsherman** [comment]
> @mattsherman that's curious... I literally just launch a JN site from this PR page,
> 
> * skip the core profiler,
> * set my country to US:CA,
> * enable product editor beta
> * go to add product, variations
> * add variation
> * see the block error
> 
> enjoy your AFK, let's look into this next week 😬

Okay, I don't know why this wasn't appearing for me when testing #49117 with JN before, but with that PR, or this rebased PR, I see the issue in JN. Will dig in.

**Comment by mattsherman** [comment]
@rjchow I think I really have it this time. I have the following PR up for review by Mothra:

- https://github.com/woocommerce/woocommerce/pull/49248

I've also tested those change with the changes from this PR, in the branch for #49216 

I don't get the crash anymore on JN or locally with React 18. But, I thought that was the case previously (still don't know how I managed to think it was fixed when it wasn't?!).

I provide an update there when that PR is merged.

**Comment by mattsherman** [comment]
> @rjchow I think I really have it this time. I have the following PR up for review by Mothra:
> 
> * [Product Editor: Fix crash and issues with variations (that will cause problems with React 18) #49248](https://github.com/woocommerce/woocommerce/pull/49248)
> 
> I've also tested those change with the changes from this PR, in the branch for #49216
> 
> I don't get the crash anymore on JN or locally with React 18. But, I thought that was the case previously (still don't know how I managed to think it was fixed when it wasn't?!).
> 
> I provide an update there when that PR is merged.

This PR has been reviewed and merged. The underlying error still occurs, but it is suppressed.

**Comment by louwie17** [comment]
@rjchow I rebased this, I am also just doing a round of testing, should finish up tomorrow morning.

**Review by louwie17 — APPROVED** [review]
This tested well to me now with the additional fixes Matt added.
If we wanted we could wait to merge this until next week Wednesday, so it lives in trunk a little longer before code freeze, but no hard opinion on that.

**Comment by louwie17** [comment]
> > @rjchow I think I really have it this time. I have the following PR up for review by Mothra:
> > 
> > * [Product Editor: Fix crash and issues with variations (that will cause problems with React 18) #49248](https://github.com/woocommerce/woocommerce/pull/49248)
> > 
> > I've also tested those change with the changes from this PR, in the branch for #49216
> > I don't get the crash anymore on JN or locally with React 18. But, I thought that was the case previously (still don't know how I managed to think it was fixed when it wasn't?!).
> > I provide an update there when that PR is merged.
> 
> This PR has been reviewed and merged. The underlying error still occurs, but it is suppressed.

@mattsherman  I figured out why the error was happening, because of a couple things:
- The error happened because the `'woocommerce/product-tab'` block wasn't registered anymore, but shortly after was again. The code relied on the data store to find the block, but couldn't find it.
- The error only happened when creating variations on a brand new product

We have some code in place that registers and unregisters the blocks, we technically only want to do this when the actual edit product page is being unmounted ( see relevant code: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/products/product-page.tsx#L97 ).
But at some point in the past we added the `productId` to the dependency array. This caused the blocks to be unregistered/registered again when the editor saves the product and redirects to the new route with the product id.
This clearly caused a timing issue for when blocks where being re-rendered and the store re-registered.
Moving the `registerBlocks`/`unregisterBlocks` to its own `useEffect` without a dependency fixes this.

**Comment by louwie17** [comment]
@rjchow do you think we can re-base this and merge it now?

**Comment by rjchow** [comment]
> @rjchow do you think we can re-base this and merge it now?

Hey @louwie17 thanks for the reminder! I was going to circle back around to this after 9.2 code freeze - which has happened today. I'll test it out later today and confirm that all's good!

**Comment by rjchow** [comment]
Tests well, no more errors 🎉 
Thanks for all the hard work @mattsherman and @louwie17 !

**Comment by rjchow** [comment]
Still seeing some e2e test failures that look consistent, will investigate this tomorrow

**Comment by rjchow** [comment]
@louwie17 I was stepping through the [e2e test failure](https://github.com/woocommerce/woocommerce/actions/runs/10139545927/job/28033071056?pr=48785) and it seems like the execution of this line is causing a block error:
https://github.com/woocommerce/woocommerce/blob/e14764ef11160b47ff33e8b4713f8435b676e9d3/plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js#L80

It seems to be a [setState loop](https://react.dev/errors/185?invariant=185) or something like that. The really strange thing is that it is only triggered if the playwright selector.click() is executed. It doesn't seem to be triggered if I step through the debugger until just before the click(), and then I click it manually in the playwright session, and continue with the debug. I suspect there could be some intricate race condition that is only triggered by the playwright click(). 

I tried putting a 5s delay after typing in the product name but it didn't fix it. I also tried getting the selector to click on the grandparent \<li\> instead, but it also didn't seem to resolve the issue. 

There also seems to be a related test failure caused by an unexpected re-render and reset of state in this test suite: https://github.com/woocommerce/woocommerce/actions/runs/10139545927/job/28033070899?pr=48785 . I could not reproduce this locally in both e2e and in regular UI, however

Could you have a look at this please?

**Comment by louwie17** [comment]
> @louwie17 I was stepping through the [e2e test failure](https://github.com/woocommerce/woocommerce/actions/runs/10139545927/job/28033071056?pr=48785) and it seems like the execution of this line is causing a block error:
> 
> https://github.com/woocommerce/woocommerce/blob/e14764ef11160b47ff33e8b4713f8435b676e9d3/plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js#L80
> 
> It seems to be a [setState loop](https://react.dev/errors/185?invariant=185) or something like that. The really strange thing is that it is only triggered if the playwright selector.click() is executed. It doesn't seem to be triggered if I step through the debugger until just before the click(), and then I click it manually in the playwright session, and continue with the debug. I suspect there could be some intricate race condition that is only triggered by the playwright click().
> 
> I tried putting a 5s delay after typing in the product name but it didn't fix it. I also tried getting the selector to click on the grandparent <li> instead, but it also didn't seem to resolve the issue.
> 
> There also seems to be a related test failure caused by an unexpected re-render and reset of state in this test suite: https://github.com/woocommerce/woocommerce/actions/runs/10139545927/job/28033070899?pr=48785 . I could not reproduce this locally in both e2e and in regular UI, however
> 
> Could you have a look at this please?

Nice sleuthing, I will look into the first issue first. It seems like the linked products logic has some `useEffect` nightmares, which is causing the above loop. I will have to do a bit of refactoring to fix this.

**Comment by louwie17** [comment]
@rjchow quick update, I do have a [PR ready](https://github.com/woocommerce/woocommerce/pull/50128) that fixes the E2E issue, I will rebase this branch when that is merged.

**Comment by louwie17** [comment]
> There also seems to be a related test failure caused by an unexpected re-render and reset of state in this test suite: https://github.com/woocommerce/woocommerce/actions/runs/10139545927/job/28033070899?pr=48785 . I could not reproduce this locally in both e2e and in regular UI, however

To follow up on these, from looking at the video it looked like it targeted the wrong select controls causing failures. Not sure why this was not exactly happening locally, but I updated the targeting to make sure it targets the correct dropdowns in this commit: https://github.com/woocommerce/woocommerce/pull/48785/commits/172f88abad3f2ead4fea8e7b80e8546471770f07 I hope that fixes those E2E failures.

The tests were relying on an select control ID that included an `index`, given React 18 enables concurrent mode I am guessing that these indexes are less consistent because one select control could be rendered earlier than another?

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/customer-list.spec.js
Comment (louwie17) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/customer-list.spec.js:122]: These tests were targeting the select control input by `id` which included a generated `index`, this seemed like a bad idea, especially if React 18 supports concurrent mode, meaning these indexes could depend on timing.

Comment (rjchow) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/customer-list.spec.js:122]: Ah ok, thanks for that! I wasn't aware that those were generated indexes

**Comment by louwie17** [comment]
✅ ✅ ✅ 🎉

**Comment by rjchow** [comment]
Oh it auto-merged from the last time I flagged that 😆 

Thanks so much @louwie17 !!

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/customer-list.spec.js
Comment (louwie17) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/customer-list.spec.js:122]: I only figured it out by looking at the video of the failed test and comparing it to my local version. Noticing it was adding the email to the wrong select control in the failing test.

**Pull request merged**
